synapse/cortex.py

```diff
@@ -743,38 +873,179 @@ def addRuntPropDel(self, prop, func):
+    #async def setLayerInfo(
+
+    def getLayer(self, iden=None):
+        if iden is None:
+            iden = self.iden
+        return self.layers.get(iden)
+
+    def getView(self, iden=None):
+        if iden is None:
+            iden = self.iden
+        return self.views.get(iden)
+
+    async def addLayer(self, **info):
+        '''
+        info = {
+            'iden': <str>, ( optional iden. default guid() )
+            'type': <str>, ( optional type. default lmdb )
+            'owner': <str>, ( optional owner. default root )
+            'config': {}, # type specific config options.
+        }
+        '''
+        iden = info.pop('iden', None)
+        if iden is None:
+            iden = s_common.guid()
+
+        node = await self.hive.open(('cortex', 'layers', iden))
+
+        layrinfo = await node.dict()
+        layrconf = await (await node.open('config')).dict()
+
+        await layrinfo.set('type', info.get('type', 'lmdb'))
+        await layrinfo.set('owner', info.get('owner', 'root'))
+        await layrinfo.set('name', info.get('name', '??'))
+
+        for name, valu in info.get('config', {}).items():
+            await layrconf.set(name, valu)
+
+        return await self._layrFromNode(node)
+
+    async def joinTeleLayer(self, url, indx=0):
+        '''
+        Convenience function to join a remote telepath layer
+        into this cortex and default view.
+        '''
+        info = {
+            'type': 'remote',
+            'owner': 'root',
+            'config': {
+                'url': url
+            }
+        }
+
+        layr = await self.addLayer(**info)
+        self.view.addLayer(layr)
+        return layr.iden
+
+    async def _layrFromNode(self, node):
+
+        info = await node.dict()
+
+        ctor = self.layrctors.get(info.get('type'))
+        if ctor is None:
+            logger.warning('layer has invalid type: %r %r' % (node.name(), info.get('type')))
+            return None
+
+        layr = await ctor.anit(self, node)
+        self.layers[layr.iden] = layr
+
+        return layr
+
     async def _initCoreLayers(self):
 
-        import synapse.cells as s_cells  # avoid import cycle
+        #import synapse.cells as s_cells  # avoid import cycle
 
-        layersdir = pathlib.Path(self.dirn, 'layers')
-        layersdir.mkdir(exist_ok=True)
+        dirn = s_common.gendir(self.dirn, 'layers')
+        node = await self.hive.open(('cortex', 'layers'))
 
-        # Layers are imported in reverse lexicographic order, where the earliest in the alphabet is the 'topmost'
-        # write layer.
-        layerdirs = sorted((d for d in layersdir.iterdir() if d.is_dir()), reverse=True)
-        for layeridx, layerdir in enumerate(layerdirs):
+        # TODO eventually hold this and watch for changes
+        for iden, node in node:
+            await self._layrFromNode(node)
 
-            logger.info('loading external layer from %s', layerdir)
+        self._migrOrigLayer()
```

Comment:
currently deployed other layers are the weird readonly / lmdb crazy docker mapped volume stuff that can't be de-facto ported.  since nobody else is likely to have any, i figured this was OK until release :)  Thoughts?